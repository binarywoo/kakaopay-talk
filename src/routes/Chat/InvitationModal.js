import React from 'react'
import { Modal, List } from 'antd'
import moment from 'moment'
import SpinContainer from '../../components/SpinContainer/index'

const InvitationModal = ({
  show,
  userList,
  chat,
  user,
  onClose,
  onClickUser
}) => {
  return (
    <Modal
      visible={show}
      title='대화상대 초대'
      footer={null}
      onCancel={onClose}
    >
      {!userList && <SpinContainer height='30vh' />}
      {userList && (
        <List
          style={{
            padding: '0 12px'
          }}
          itemLayout='horizontal'
          dataSource={userList.filter(
            item => item.key !== user.key && !chat.users[item.key]
          )}
          renderItem={user => {
            moment.locale('ko')
            return (
              <List.Item
                onClick={() => onClickUser(user)}
                style={{ cursor: 'pointer' }}
              >
                <List.Item.Meta title={user.userId} />
              </List.Item>
            )
          }}
        />
      )}
    </Modal>
  )
}

export default InvitationModal
