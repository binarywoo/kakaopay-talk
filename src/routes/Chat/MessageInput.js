import React, { useCallback } from 'react'
import {
  Avatar,
  Upload,
  Row,
  Col,
  Input,
  Icon,
  message as antdMessage
} from 'antd'
import classNames from 'classnames'
import MessageModal from '../../components/MessageModal/index'

const MessageInput = ({
  message,
  onChange,
  onSendMessage,
  onClickInvitation,
  sendImageMessage,
  showHiddenMenu,
  setShowHiddenMenu
}) => {
  const beforeImageUpload = useCallback(file => {
    setShowHiddenMenu(false)
    const hide = antdMessage.loading('이미지 업로드 중..', 0)
    sendImageMessage(file).then(() => hide())
    return false
  }, [])

  const onClickNotReady = useCallback(() => {
    MessageModal({
      type: 'info',
      title: '준비중',
      content: '준비중인 기능입니다.'
    })
  }, [])

  return (
    <div className='chat-input-wrapper' offsetBottom={0}>
      <div
        style={{
          padding: 12,
          backgroundColor: '#fff',
          borderTop: '1px solid #bbbbbb'
        }}
      >
        <Input.Search
          value={message}
          onChange={onChange}
          addonBefore={
            <Icon
              type={showHiddenMenu ? 'minus' : 'plus'}
              onClick={() => setShowHiddenMenu(!showHiddenMenu)}
            />
          }
          enterButton={<Icon type='arrow-up' />}
          onSearch={message.length > 0 ? onSendMessage : null}
        />
      </div>
      <div
        className={classNames(
          'hidden-menu',
          showHiddenMenu ? 'show' : 'hidden'
        )}
      >
        <Row gutter={12}>
          <Col span={6}>
            <Avatar
              className='invitation'
              icon='user-add'
              onClick={onClickInvitation}
            />
            초대
          </Col>
          <Col span={6}>
            <Upload
              key='picture'
              accept='image/*'
              beforeUpload={beforeImageUpload}
            >
              <Avatar className='image' icon='picture' />
              이미지
            </Upload>
          </Col>
          <Col span={6}>
            <Avatar className='pay' icon='wallet' onClick={onClickNotReady} />
            송금
          </Col>
          <Col span={6}>
            <Avatar
              className='map'
              icon='environment'
              onClick={onClickNotReady}
            />
            지도
          </Col>
        </Row>
      </div>
    </div>
  )
}

export default MessageInput
