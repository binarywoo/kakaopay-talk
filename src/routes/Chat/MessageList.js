import React from 'react'
import moment from 'moment'
import Message from '../../components/Message/index'

const MessageList = ({ messages, user }) => {
  return (
    <div>
      {messages &&
        messages.map((item, idx) => {
          const isMyMessage = user.key === item.user
          const time = moment(item.lastUpdate).format('HH:mm')
          const prevMessage = idx > 0 ? messages[idx - 1] : null
          const prevDate = prevMessage
            ? moment(prevMessage.lastUpdate).format('YYYY-MM-DD')
            : null
          const date = moment(item.lastUpdate).format('YYYY-MM-DD')
          const nextMessage =
            idx < messages.length - 1 ? messages[idx + 1] : null
          const nextTime = nextMessage
            ? moment(nextMessage.lastUpdate).format('HH:mm')
            : null
          const nextDate = nextMessage
            ? moment(nextMessage.lastUpdate).format('YYYY-MM-DD')
            : null
          const nextUserId =
            idx < messages.length - 1 ? nextMessage.userId : null
          const prevUserId = prevMessage ? prevMessage.userId : null
          const prevTime = prevMessage
            ? moment(prevMessage.lastUpdate).format('HH:mm')
            : null

          let showTag = true
          let showDivider = false
          let showProfile = false
          if (
            nextTime === time &&
            nextDate === date &&
            nextUserId === item.userId
          )
            showTag = false
          if (prevDate !== date) showDivider = true
          if (
            item.userId !== prevUserId ||
            prevTime !== time ||
            prevDate !== date
          )
            showProfile = true
          return (
            <Message
              isMyMessage={isMyMessage}
              showDivider={showDivider}
              showTag={showTag}
              showProfile={showProfile}
              date={date}
              time={time}
              item={item}
            />
          )
        })}
    </div>
  )
}

export default MessageList
