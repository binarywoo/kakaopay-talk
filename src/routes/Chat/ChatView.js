import React, { useEffect, useState, useCallback } from 'react'
import { Modal } from 'antd'

import ContentContainer from '../../components/ContentContainer/index'
import './index.less'
import { scrollToBottom } from '../../libs/chatAppUtils'
import MessageList from './MessageList'
import MessageInput from './MessageInput'
import InvitationModal from './InvitationModal'

const ChatView = ({
  messages,
  chat,
  user,
  onSendMessage,
  getUserListToInvite,
  sendInvitation,
  sendImageMessage
}) => {
  const [message, setMessage] = useState('')
  const [showUserSearchModal, setShowUserSearchModal] = useState(false)
  const [userList, setUserList] = useState(null)
  const [showHiddenMenu, setShowHiddenMenu] = useState(false)

  useEffect(() => {
    scrollToBottom()
  }, [messages])

  useEffect(() => {
    scrollToBottom()
  }, [showHiddenMenu])

  const handleOnSendMessage = useCallback(() => {
    onSendMessage(message)
    setMessage('')
  }, [message])

  const onClickInvitation = useCallback(() => {
    setShowHiddenMenu(false)
    getUserListToInvite().then(userList => {
      setUserList(userList)
    })
    setShowUserSearchModal(true)
  }, [])

  const onClickUser = useCallback(
    userToInvite => {
      const newInvitation = {
        from: user.key,
        fromId: user.userId,
        to: userToInvite.key,
        toId: userToInvite.userId,
        chat: chat.key,
        chatTitle: chat.title,
        lastupdate: new Date().toISOString()
      }
      sendInvitation(newInvitation).then(() => {
        Modal.success({
          title: '초대 성공',
          content: `${userToInvite.userId}님께 초대메시지를 보냈습니다.`,
          okText: '확인',
          onOk: () => {
            setShowUserSearchModal(false)
          }
        })
      })
    },
    [userList]
  )

  return (
    <ContentContainer
      style={{ background: user.backgroundColor || 'rgb(18, 115, 222)' }}
    >
      <div
        className={!showHiddenMenu ? 'smooth-padding' : null}
        style={{ padding: `12px 12px ${showHiddenMenu ? 168 : 68}px 12px` }}
      >
        <MessageList messages={messages} user={user} />
      </div>
      <MessageInput
        message={message}
        onChange={e => setMessage(e.target.value)}
        onSendMessage={handleOnSendMessage}
        onClickInvitation={onClickInvitation}
        showHiddenMenu={showHiddenMenu}
        setShowHiddenMenu={setShowHiddenMenu}
        sendImageMessage={sendImageMessage}
      />
      <InvitationModal
        show={showUserSearchModal}
        userList={userList}
        chat={chat}
        user={user}
        onClose={() => setShowUserSearchModal(false)}
        onClickUser={onClickUser}
      />
    </ContentContainer>
  )
}

export default ChatView
