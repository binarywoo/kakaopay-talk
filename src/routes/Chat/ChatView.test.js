import React from 'react'
import { shallow } from 'enzyme'
import ChatView from './ChatView'
import MessageList from './MessageList'
import MessageInput from './MessageInput'
import InvitationModal from './InvitationModal'
import {
  DUMMY_MESSAGE,
  DUMMY_IMAGE_MESSAGE,
  DUMMY_CHAT,
  DUMMY_USER
} from '../../constants/testConstants'

describe('ChatView', () => {
  let wrapper = null

  const props = {
    messages: [DUMMY_MESSAGE, DUMMY_IMAGE_MESSAGE],
    chat: DUMMY_CHAT,
    user: DUMMY_USER,
    onSendMessage: () => null,
    getUserListToInvite: () => null,
    sendInvitation: () => null,
    sendImageMessage: () => null
  }

  it('정상적으로 렌더링이 된다.', () => {
    wrapper = shallow(<ChatView {...props} />)
  })

  it('스냅샷과 매칭이 된다.', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('하위 컴포넌트들을 정상적으로 렌더링한다.', () => {
    expect(wrapper.find(MessageList).length).toBe(1)
    expect(wrapper.find(MessageInput).length).toBe(1)
    expect(wrapper.find(InvitationModal).length).toBe(1)
  })
})
