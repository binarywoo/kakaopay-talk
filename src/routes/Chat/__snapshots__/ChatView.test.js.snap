// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatView 스냅샷과 매칭이 된다. 1`] = `
<ContentContainer
  style={
    Object {
      "background": "rgb(18, 115, 222)",
    }
  }
>
  <div
    className="smooth-padding"
    style={
      Object {
        "padding": "12px 12px 68px 12px",
      }
    }
  >
    <MessageList
      messages={
        Array [
          Object {
            "content": "하이",
            "key": "-LbnOJfNoldq3SD0xRIR",
            "lastUpdate": "2019-04-06T16:04:55.757Z",
            "type": "text",
            "user": "-LbmTnQw-OjNS87fDbBt",
            "userId": "이웅희",
          },
          Object {
            "content": "하이",
            "key": "-LbnOJfNolds3SD0xRIR",
            "lastUpdate": "2019-04-06T16:06:55.757Z",
            "type": "image",
            "user": "-LbmTnQw-OjNS87fDbBt",
            "userId": "이웅희",
          },
        ]
      }
      user={
        Object {
          "key": "-LbmTnQw-OjNS87fDbBt",
          "userId": "이웅희",
        }
      }
    />
  </div>
  <MessageInput
    message=""
    onChange={[Function]}
    onClickInvitation={[Function]}
    onSendMessage={[Function]}
    sendImageMessage={[Function]}
    setShowHiddenMenu={[Function]}
    showHiddenMenu={false}
  />
  <InvitationModal
    chat={
      Object {
        "key": "testKey",
        "lastMessage": "test message",
        "lastUpdate": "2019-04-06T18:32:11.433Z",
        "title": "test chat",
      }
    }
    onClickUser={[Function]}
    onClose={[Function]}
    show={false}
    user={
      Object {
        "key": "-LbmTnQw-OjNS87fDbBt",
        "userId": "이웅희",
      }
    }
    userList={null}
  />
</ContentContainer>
`;
